lxd/network/network_utils.go

```diff
@@ -639,6 +639,109 @@ func randomSubnetV6() (string, error) {
 	return "", fmt.Errorf("Failed to automatically find an unused IPv6 subnet, manual configuration required")
 }
 
+// randomAddressInSubnet finds a random address within a given subnet. If given, the validate function is called for
+// each randomly generated value. Each time validate returns an error, another IP address will be generated until the
+// context is cancelled.
+func randomAddressInSubnet(ctx context.Context, ipnet *net.IPNet, validate func(net.IP) error) (net.IP, error) {
+	var bufferLen int
+	var ipVersion int
+	if ipnet.IP.To4() != nil {
+		ipVersion = 4
+		bufferLen = 4
+	} else {
+		ipVersion = 6
+		bufferLen = 16
+	}
+
+	ones, _ := ipnet.Mask.Size()
+
+	// The number of bytes that will be the same in the subnet and in the random address.
+	quotient := ones / 8
+
+	// The remaining bits that will be the same in the subnet and in the random address (after the equivalent bytes).
+	remainder := ones % 8
+
+	// Can't get a random address from e.g. 192.0.2.1/32 as it's a single IP.
+	if quotient == bufferLen {
+		return nil, fmt.Errorf("Cannot find random address when subnet contains only one address")
```

Comment:
I would have expected this function to always return the same IP if a single /32 or /128 IP is passed in, rather than error.

Reply:
I suppose this function should be keeping track of the offered IPs to the validate function and will not re-offer the same one repeatedly. And when there are no IPs left to offer fail out of the validate function hasn't returned nil by then.

Reply:
Basically a single host value shouldnt be a special case :)

Reply:
> I would have expected this function to always return the same IP if a single /32 or /128 IP is passed in, rather than error.
> Basically a single host value shouldnt be a special case - for the caller :)

In this case the only IP that can be returned would be the network address of the subnet itself, which is skipped below. This error essentially shortcuts the call so that we don't have to wait for the timeout.

> I suppose this function should be keeping track of the offered IPs to the validate function and will not re-offer the same one repeatedly. And when there are no IPs left to offer fail out of the validate function hasn't returned nil by then.

In my first iteration of this method I was figuring out the number of available IPs first, then picking a random number and mapping that back to the subnet. It turned out to be very complicated and I couldn't get it to pass my own unit tests :cry: 

Depending on the size of the subnet, keeping track of the IPs that have been attempted gets complicated quickly. E.g. if we have a map of attempted IPs, you could (very theoretically) have a map of size 2^128. In the end the simplest thing was to just generate random IPs until a timeout is reached.